### PR TITLE
[SPARK-25421][SQL] Abstract an output path field in trait DataWritingCommand

### PR DESCRIPTION
**Pull request by LantaoJin**
## What changes were proposed in this pull request?

#22353 import a metadata field in ```SparkPlanInfo``` and it could dump the input location for read. Corresponding, we need add a field in ```DataWritingCommand``` for output path.

## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by LantaoJin** [comment]
Gently ping @cloud-fan @dongjoon-hyun , would you please help to review?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -440,7 +440,7 @@ case class DataSource(
     // ordering of data.logicalPlan (partition columns are all moved after data column).  This
     // will be adjusted within InsertIntoHadoopFsRelation.
     InsertIntoHadoopFsRelationCommand(
-      outputPath = outputPath,
+      outputFsPath = outputPath,
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Could you undo this redundant change, @LantaoJin ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala [hunk 2]
```diff
@@ -18,6 +18,7 @@
 package org.apache.spark.sql.execution
 
 import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.sql.execution.command.{DataWritingCommand, DataWritingCommandExec}
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala]: Do we need `DataWritingCommand` here, too?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 1 above)
Comment (LantaoJin) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: This field overwrites the outputPath in DataWritingCommand and the return type is different (Path vs Option[Path]), so I rename this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala (hunk 2 above)
Comment (LantaoJin) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala]: Will remove it

**Comment by LantaoJin** [comment]
Gently ping @dongjoon-hyun @cloud-fan

**Comment by cloud-fan** [comment]
This time it's not a regression right? I'd like to not change the interface, but explicitly catch the plan we are interested in `fromSparkPlan`, e.g.
```
case DataWritingCommandExec(i: InsertIntoHadoopFsRelation, _) =>
  Map("path" -> i.outputPath)
```

**Comment by LantaoJin** [comment]
Isn't common? I am afraid not only one InsertIntoHadoopFsRelation need to added in case statment.

**Comment by cloud-fan** [comment]
Without a thorough design, I hesitate to change the `DataWritingCommand` interface only for event log. Do you have any more plans to improve the event log?

**Comment by LantaoJin** [comment]
Most of the information we wanted could be analyzed out from event log except some metrics in Executor side which doesn't heartbeat to Driver, e.g RPC count with NameNode. Another case is #21221, before that we had to hack code to get the similar metrics. Event log as a structured, unified, overall, replay-able log, it offers a possibility to analysis offline, even realtime. We prefer to use it since the history UI exposes less information than user expected, further more not smart and hard to customize. We are on going on this based on event log. Thanks @cloud-fan,  I suggest to add this interface to ```DataWritingCommand```. Pattern matching each implementations looks trick. It looks common, maybe it could be used in physical plan optimization in future.

**Comment by cloud-fan** [comment]
Since this is a new feature, we can't just merge it like #22353 without a proper design.

Making the event logs as a structured, unified and reliable source for Spark metrics looks like a good idea. Let's write a design doc to explain what we already have in the event logs, and what is missing, and how to make it reliable, and what's the issue if we read it in real time. It's better to discuss it in the dev list and see if other people have different ideas to get Spark metrics.

**Comment by LantaoJin** [comment]
Agree that. Since this field is important to us. Could I refactor it following your advice and file a discussion in another Jira?

**Comment by LantaoJin** [comment]
Using pattern matching will face a problem. ```InsertIntoHiveDirCommand```,```CreateHiveTableAsSelectCommand``` and ```InsertIntoHiveTable``` are all in spark-hive module. SparkPlanInfo could not include them.

**Comment by LantaoJin** [comment]
@cloud-fan I refactor and remove the function outputPath in ```DataWritingCommand```. Besides the unit test you could see, in my local, I added below test in ```HiveQuerySuite.scala```:
```scala
  test("SPARK-25421 DataWritingCommandExec(hive) should contains 'OutputPath' metadata") {
    withTable("t") {
      sql("CREATE TABLE t(col_I int)")
      val f = sql("INSERT OVERWRITE TABLE t SELECT 1")
      assert(SparkPlanInfo.fromSparkPlan(f.queryExecution.sparkPlan).metadata
        .contains("OutputPath"))
    }
  }
```
But since ```HiveQuerySuite``` can not access ```SparkPlanInfo```, after test passed in my local, I remove it again.

**Comment by LantaoJin** [comment]
Gently ping @cloud-fan

**Pull request closed without merging**